src/Functions/LeastGreatestGeneric.h

```diff
@@ -114,6 +125,24 @@ class LeastGreatestOverloadResolver : public IFunctionOverloadResolver
         if (arguments.size() == 2 && isNumber(removeNullable(arguments[0].type)) && isNumber(removeNullable(arguments[1].type)))
             return std::make_unique<FunctionToFunctionBaseAdaptor>(SpecializedFunction::create(context), argument_types, return_type);
 
+        /// In sql query like `least(x, NULL)` or `greatest(x, NULL)`, which has constant NULL arguments, will make the return type is only null.
+        /// In such case, we should reconstruct the return type based on the not only null input arguments.
+        if (return_type->onlyNull())
+        {
+            DataTypes not_only_null_types;
+            for (const auto & argument_type : argument_types)
+            {
+                if (!argument_type->onlyNull())
+                    not_only_null_types.push_back(argument_type);
+            }
+            if (!not_only_null_types.empty())
+            {
+                const DataTypePtr & result_type = removeLowCardinality(getLeastSupertype(not_only_null_types));
+                return std::make_unique<FunctionToFunctionBaseAdaptor>(
```

Comment:
This copy-pastes l. 146. Can we instead please add 'DateTypePtr return_type_copy = return_type;' after l. 120 and only modify 'return_type_copy' in l. 140, delete l. 141, and change l. 146 from 'return_type' to 'return_type_copy'?